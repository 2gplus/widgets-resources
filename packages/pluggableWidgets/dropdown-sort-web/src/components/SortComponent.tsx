import { createElement, ReactElement, useCallback, useEffect, useRef, useState } from "react";
import { SortDirection, useOnClickOutside } from "@mendix/piw-utils-internal";
import classNames from "classnames";

export interface SortOption {
    caption: string;
    value: string;
}

interface SortComponentProps {
    ariaLabel?: string;
    defaultDirection?: SortDirection;
    defaultOption?: SortOption;
    emptyOptionCaption?: string;
    name?: string;
    options: SortOption[];
    tabIndex?: number;
    updateSort?: (value: SortOption, direction: SortDirection) => void;
}

export function SortComponent(props: SortComponentProps): ReactElement {
    const [valueInput, setValueInput] = useState(props.defaultOption?.caption ?? props.emptyOptionCaption ?? "");
    const [selectedSort, setSelectedSort] = useState<SortOption>(
        props.defaultOption ?? {
            caption: props.emptyOptionCaption ?? "",
            value: ""
        }
    );
    const [show, setShow] = useState(false);
    const [dropdownWidth, setDropdownWidth] = useState(0);
    const [direction, setDirection] = useState(props.defaultDirection ?? "asc");

    const componentRef = useRef<HTMLDivElement>(null);

    const onClick = useCallback((option: SortOption) => {
        setValueInput(option.caption);
        setSelectedSort(option);
        setShow(false);
    }, []);

    useOnClickOutside(componentRef, () => setShow(false));

    useEffect(() => {
        if (selectedSort) {
            props.updateSort?.(selectedSort, direction);
        }
    }, [direction, selectedSort]);

    const showPlaceholder = !selectedSort || valueInput === props.emptyOptionCaption;

    return (
        <div className="dropdown-container" data-focusindex={props.tabIndex ?? 0} ref={componentRef}>
            <div className="dropdown-triggerer-wrapper">
                <input
                    value={showPlaceholder ? "" : valueInput}
                    placeholder={showPlaceholder ? props.emptyOptionCaption : undefined}
                    className="form-control dropdown-triggerer"
                    onClick={() => setShow(true)}
                    onFocus={() => setShow(true)}
                    aria-haspopup
                    ref={inputRef => {
                        if (inputRef && inputRef.clientWidth) {
                            setDropdownWidth(inputRef.clientWidth);
                        }
                    }}
                    aria-expanded={show}
                    aria-controls={`${props.name}-dropdown-list`}
                    aria-label={props.ariaLabel}
                />
                <button
                    className={classNames("btn btn-default btn-sort", {
                        "icon-asc": direction === "asc",
                        "icon-desc": direction === "desc"
                    })}
                    onClick={() => setDirection(prev => (prev === "asc" ? "desc" : "asc"))}
                />
            </div>
            {show && (
                <ul
                    id={`${props.name}-dropdown-list`}
                    className="dropdown-list"
                    style={{ width: dropdownWidth }}
                    role="menu"
                    data-focusindex={0}
                >
                    {props.options.map((option, index) => (
                        <li
                            className={classNames({
                                "filter-selected": selectedSort?.value === option.value
                            })}
                            key={index}
                            onClick={() => onClick(option)}
                            onKeyDown={e => {
                                if (e.key === "Enter" || e.key === " ") {
                                    e.preventDefault();
                                    onClick(option);
                                }
                            }}
                            role="menuitem"
                            tabIndex={0}
                        >
                            <div className="filter-label">{option.caption}</div>
                        </li>
                    ))}
                </ul>
            )}
        </div>
    );
}
