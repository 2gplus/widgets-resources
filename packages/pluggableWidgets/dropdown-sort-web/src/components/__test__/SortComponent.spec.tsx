import { render, shallow } from "enzyme";
import { createElement } from "react";
import { render as renderTestingLib, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import { SortComponent, SortOption } from "../SortComponent";

const defaultOptions: SortOption[] = [
    { caption: "1", value: "_1" },
    { caption: "2", value: "_2" },
    { caption: "3", value: "_3" }
];

jest.useFakeTimers();

describe("Sort selector", () => {
    describe("renders correctly", () => {
        it("with options", () => {
            const component = render(<SortComponent options={defaultOptions} />);

            expect(component).toMatchSnapshot();
        });
        it("with no options", () => {
            const component = render(<SortComponent options={[]} />);

            expect(component).toMatchSnapshot();
        });
        it("with ariaLabel", () => {
            const component = render(<SortComponent options={defaultOptions} ariaLabel="my label" />);

            expect(component).toMatchSnapshot();
        });
        it("with emptyOptioncaption", () => {
            const component = render(<SortComponent options={defaultOptions} emptyOptionCaption={"find me"} />);

            expect(component).toMatchSnapshot();
            expect(component.find("input").first().prop("placeholder")).toBe("find me");
        });
    });
    it("selects default option", () => {
        const defaultOption = defaultOptions[0];
        const component = shallow(<SortComponent options={defaultOptions} defaultOption={defaultOption} />);

        const input = component.find("input").first();

        expect(input.prop("value")).toBe(defaultOption.caption);
    });

    describe("when value changes", () => {
        it("calls updateFilters when value changes", () => {
            const onClickProps = { preventDefault: jest.fn(), stopPropagation: jest.fn() };
            const updateSortHandler = jest.fn();
            const component = shallow(<SortComponent options={defaultOptions} updateSort={updateSortHandler} />);

            const input = component.find("input");
            input.simulate("click", onClickProps);

            const item = component.find("li").first();
            item.simulate("click", onClickProps);

            expect(updateSortHandler).toBeCalled();
        });
        it("shows selected option on input value", () => {
            const onClickProps = { preventDefault: jest.fn(), stopPropagation: jest.fn() };
            const defaultOption = defaultOptions[1];
            const component = shallow(<SortComponent options={defaultOptions} defaultOption={defaultOption} />);

            const input = component.find("input");
            input.simulate("click", onClickProps);

            expect(component.find("input").first().prop("value")).toBe(defaultOption.caption);

            const item = component.find("li").last(); // [cap 3: val:_3]
            item.simulate("click", onClickProps);

            expect(component.find("input").first().prop("value")).toBe(defaultOptions[2].caption);
        });
    });

    describe("focus", () => {
        beforeEach(() => (document.body.innerHTML = ""));

        it("changes focused element when pressing the input", () => {
            const component = renderTestingLib(
                <SortComponent options={defaultOptions} emptyOptionCaption="Click me" />
            );

            expect(document.body).toHaveFocus();
            const input = component.getByPlaceholderText("Click me");
            expect(input).toBeDefined();
            fireEvent.click(input);

            jest.advanceTimersByTime(10);

            const items = component.getAllByRole("menuitem");
            expect(items[0]).toHaveFocus();
        });

        it("changes focused element back to the input when pressing shift+tab in the first element", () => {
            const component = renderTestingLib(
                <SortComponent options={defaultOptions} emptyOptionCaption="Click me" />
            );

            expect(document.body).toHaveFocus();

            const input = component.getByPlaceholderText("Click me");
            expect(input).toBeDefined();
            fireEvent.click(input);

            jest.advanceTimersByTime(10);

            const items = component.getAllByRole("menuitem");
            expect(items[0]).toHaveFocus();

            userEvent.tab({ shift: true });

            jest.advanceTimersByTime(10);

            expect(input).toHaveFocus();
        });

        it("changes focused element back to the input when pressing tab on the last item", () => {
            const component = renderTestingLib(
                <SortComponent options={[{ caption: "1", value: "_1" }]} emptyOptionCaption="Click me" />
            );

            expect(document.body).toHaveFocus();

            const input = component.getByPlaceholderText("Click me");
            fireEvent.click(input);

            jest.advanceTimersByTime(10);

            const items = component.getAllByRole("menuitem");
            expect(items[0]).toHaveFocus();

            userEvent.tab();
            expect(items[1]).toHaveFocus();
            userEvent.tab();

            jest.advanceTimersByTime(10);

            expect(input).toHaveFocus();
        });

        it("changes focused element back to the input when pressing escape on the last item", () => {
            const component = renderTestingLib(
                <SortComponent options={[{ caption: "1", value: "_1" }]} emptyOptionCaption="Click me" />
            );

            expect(document.body).toHaveFocus();

            const input = component.getByPlaceholderText("Click me");
            fireEvent.click(input);

            jest.advanceTimersByTime(10);

            const items = component.getAllByRole("menuitem");
            expect(items[0]).toHaveFocus();

            userEvent.tab();

            expect(items[1]).toHaveFocus();

            userEvent.keyboard("{esc}");

            jest.advanceTimersByTime(10);

            expect(input).toHaveFocus();
        });
    });
});
