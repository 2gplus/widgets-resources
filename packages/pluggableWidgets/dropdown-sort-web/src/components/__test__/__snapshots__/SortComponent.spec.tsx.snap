// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort selector renders correctly with ariaLabel 1`] = `
<div
  class="dropdown-container"
  data-focusindex="0"
>
  <div
    class="dropdown-triggerer-wrapper"
  >
    <input
      aria-controls="undefined-dropdown-list"
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="my label"
      class="form-control dropdown-triggerer"
      value=""
    />
    <button
      class="btn btn-default btn-sort icon-asc"
    />
  </div>
</div>
`;

exports[`Sort selector renders correctly with emptyOptioncaption 1`] = `
<div
  class="dropdown-container"
  data-focusindex="0"
>
  <div
    class="dropdown-triggerer-wrapper"
  >
    <input
      aria-controls="undefined-dropdown-list"
      aria-expanded="false"
      aria-haspopup="true"
      class="form-control dropdown-triggerer"
      placeholder="find me"
      value=""
    />
    <button
      class="btn btn-default btn-sort icon-asc"
    />
  </div>
</div>
`;

exports[`Sort selector renders correctly with no options 1`] = `
<div
  class="dropdown-container"
  data-focusindex="0"
>
  <div
    class="dropdown-triggerer-wrapper"
  >
    <input
      aria-controls="undefined-dropdown-list"
      aria-expanded="false"
      aria-haspopup="true"
      class="form-control dropdown-triggerer"
      value=""
    />
    <button
      class="btn btn-default btn-sort icon-asc"
    />
  </div>
</div>
`;

exports[`Sort selector renders correctly with options 1`] = `
<div
  class="dropdown-container"
  data-focusindex="0"
>
  <div
    class="dropdown-triggerer-wrapper"
  >
    <input
      aria-controls="undefined-dropdown-list"
      aria-expanded="false"
      aria-haspopup="true"
      class="form-control dropdown-triggerer"
      value=""
    />
    <button
      class="btn btn-default btn-sort icon-asc"
    />
  </div>
</div>
`;
